main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala

```diff
@@ -597,4 +597,44 @@ object SafetyError {
          |""".stripMargin
     )
   }
+
+  /**
+    * An error raised to indicate that a function marked with the `@Tailrec` annotation
+    * has at least non-tail-recursive function call.
+    *
+    * @param loc the location of the non-tail-recursive call.
+    */
+  case class NonTailRecursiveFunction(loc: SourceLocation) extends SafetyError {
```

Comment:
Maybe also include the name of the function marked annotated with `@Tailrec` in the parameter list.